rfcs/0025-reduce-scan-state-memory-usage.md

```diff
@@ -0,0 +1,97 @@
+
+## Work Breakdown/Prioritization
+
+##### Must Do (definitely required to mitigate memory usage to acceptable levels)
+1. Implement Global Transaction Witness Cache
+2. Remove Transaction Witness Scan States Embedding
+3. Implement Full/Partial Scan State Logic
+4. Downgrade Branching Scan States in Frontier
+
+##### Might Do (extra work which will further reduce memory usage)
+1. Frontier Dead Tip Pruning
+2. Scan State Icebox
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+### Dead Frontier Tip Scan State Reconstruction Attack
+Pruning probalistically dead frontier tips opens up nodes for potential denial of service attack. Pruning scan states is seen as "safe" because there is no actual loss in local data availability as long as the node maintains the ledger mask of the previous staged ledger in the transition frontier. This is because the scan state can be restored to its full format by recomputing the transaction witnesses from the previous staged ledger's target ledger. However, it is possible for an adversary to generate a fork off of an old tip in the frontier, then broadcast that, forcing any nodes which have pruned that scan state to reconstruct the transaction witnesses for it (under the current logic). This could be prevented by either adding a better heuristic for whether or not something is worth adding to our frontier (as in, if it's such a bad block, maybe we should skip it and only download it during catchup if later there is a good chain built off of it). Alternatively, this can be mitigated by only constructing transaction witnesses for tips which are "close enough" to our best tip in terms of strength. It would be a hard attack to execute in practice, but it is theoretically possible for an adversary to wait until a series of their accounts get VRF wins in nearby slots, and then make nodes perform a whole bunch of scan state witness reconstructions at the same time. All this said, it's not clear whether this attack could cause enough work to actually take nodes offline or lag behind the network enough to do anything bad.
+
+### Incorrect Implementation of Scan State Icebox Leads to High Disk I/O
+If the scan state icebox is not implemented correctly, it could potentially greatly increase the amount of disk I/O performed during participation. As such, it is very important that this is well instrumented and tested so that it does not accidentally vastly reduce performance in order to reduce the memory usage of the scan states some.
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- can the exact gains from this be calculated before implementing these optimizations, so that we can use that as a framework for informing whether or not the optimizations were successful?
```

Comment:
I liked the analysis in the beginning and coming to the 28GB/57GB numbers. Can you include a rough analysis of memory usage when we do have the cache (maybe even with and without the extra optional optimizations)? We’d want to see these at more like <4GB with conservative estimates to know for sure that these things are fixed.

Reply:
It's pretty much impossible to do any meaningful analysis on the memory savings of the proposed optimizations without having finalized consensus parameters since we cannot compute how many tips there will be at any given point in the transition frontier. I can do the rough math using the same `(k, 2k)` assumptions during the first analysis, but I would only be able to analyze the transaction witness cache savings without branch/tip pruning logic, which is where the big savings are. If you have suggestions for how to think about how many tips we expect and what would be a good probabilistic model for reasoning about dead tips without the consensus parameters, that would be great.